### ISSUE TITLE
Binary in latest Windows release not built with WASAPI support

### ISSUE THREAD

**Issue by nseyedtalebi** [HEnquist/camilladsp#129]
First, thanks for bringing CamillaDSP into the world. 

The windows binary seems to have been built without the wasapi feature flag. When I download and run the .exe, it only shows features "alsa-backend" and "websocket". 
```
Microsoft Windows [Version 10.0.19043.1165]
(c) Microsoft Corporation. All rights reserved.

C:\Users\NiMo3>camilladsp
CamillaDSP 0.6.0
Henrik Enquist <henrik.enquist@gmail.com>
A flexible tool for processing audio

Built with features: alsa-backend, websocket

USAGE:
    camilladsp [FLAGS] [OPTIONS] <configfile>
```

There doesn't seem to be any error message if you feed it a config that refers to a WASAPI device when there's no support for it in the binary. I don't think there's anything I have running that would redirect stderr. The exit code was "-1073741819".
```
C:\Users\NiMo3>camilladsp -v camilladsp\config.yml
Aug 12 18:54:58.539 DEBG Read config file Some("camilladsp\\config.yml"), module: camilladsp
Aug 12 18:54:58.543 DEBG impulse_all.wav is relative, module: camillalib::config
Aug 12 18:54:58.544 DEBG Using impulse_all.wav found relative to config file dir, module: camillalib::config
Aug 12 18:54:58.546 DEBG impulse_all.wav is relative, module: camillalib::config
Aug 12 18:54:58.547 DEBG Using impulse_all.wav found relative to config file dir, module: camillalib::config
Aug 12 18:54:58.570 DEBG Read raw data from: '\\?\C:\Users\NiMo3\camilladsp\impulse_all.wav', format: FLOAT64LE, number of coeffs: 2097152, module: camillalib::filters
Aug 12 18:54:58.573 DEBG Read wav file '\\?\C:\Users\NiMo3\camilladsp\impulse_all.wav', format: FLOAT64LE, channel: 0 of 2, samplerate: 192000, length: 1048576, module: camillalib::filters
Aug 12 18:54:58.598 DEBG Read raw data from: '\\?\C:\Users\NiMo3\camilladsp\impulse_all.wav', format: FLOAT64LE, number of coeffs: 2097152, module: camillalib::filters
Aug 12 18:54:58.601 DEBG Read wav file '\\?\C:\Users\NiMo3\camilladsp\impulse_all.wav', format: FLOAT64LE, channel: 1 of 2, samplerate: 192000, length: 1048576, module: camillalib::filters
Aug 12 18:54:58.604 DEBG Config is valid, module: camilladsp
Aug 12 18:54:58.604 DEBG Wait for config, module: camilladsp
Aug 12 18:54:58.605 DEBG Config ready, module: camilladsp
Aug 12 18:54:58.606 DEBG Using channels [true, true], module: camilladsp
Aug 12 18:54:58.608 DEBG Build new pipeline, module: camillalib::filters
Aug 12 18:54:58.609 DEBG Build from config, module: camillalib::filters
Aug 12 18:54:58.609 DEBG Using a playback channel capacity of 3 chunks., module: camillalib::wasapidevice
Aug 12 18:54:58.610 DEBG Using a capture channel capacity of 33 buffers., module: camillalib::wasapidevice
Aug 12 18:54:58.628 DEBG Read raw data from: '\\?\C:\Users\NiMo3\camilladsp\impulse_all.wav', format: FLOAT64LE, number of coeffs: 2097152, module: camillalib::filters
Aug 12 18:54:58.632 DEBG Read wav file '\\?\C:\Users\NiMo3\camilladsp\impulse_all.wav', format: FLOAT64LE, channel: 0 of 2, samplerate: 192000, length: 1048576, module: camillalib::filters
Aug 12 18:54:58.649 DEBG Conv Conv L is using 256 segments, module: camillalib::fftconv
Aug 12 18:54:58.668 DEBG Build from config, module: camillalib::filters
Aug 12 18:54:58.694 DEBG Read raw data from: '\\?\C:\Users\NiMo3\camilladsp\impulse_all.wav', format: FLOAT64LE, number of coeffs: 2097152, module: camillalib::filters
Aug 12 18:54:58.697 DEBG Read wav file '\\?\C:\Users\NiMo3\camilladsp\impulse_all.wav', format: FLOAT64LE, channel: 1 of 2, samplerate: 192000, length: 1048576, module: camillalib::filters

C:\Users\NiMo3>
```

I will try building from source to see if it works after including the flags for WASAPI. I would guess that the Windows release is auto-generated and no one has tried to use it yet - maybe a change is needed in the CI configs somewhere?

My apologies if I'm just missing something.

**Comment by HEnquist** [comment]
Thanks for opening this issue!
There is a problem in the part that generates the "Built with features:" list. Obviously alsa-backend should not be listed in a windows build. But the missing wasapi feature is fine. I didn't see any reason to build on windows without wasapi support, so I removed the feature. Now wasapi is always included on windows. I'm planning to do the same for ALSA on Linux, and CoreAudio on macOS.

When fixing that list, I'll also make the help message show lists of all supported capture and playback types.

**Comment by nseyedtalebi** [comment]
Thanks for your reply, that makes sense. After successfully building it, reading a little more of the documentation, and checking out links, I can see that my understanding was/is faulty, not the program.  I have something wrong somewhere else (my config or my machine). The devil's in the details.

**Comment by HEnquist** [comment]
What happens, or what doesn't happen?
If you attach your config and the corresponding output (with -v) I can try to help.

**Comment by nseyedtalebi** [comment]
That would be much appreciated. I've attached a zip file with two configs, the corresponding output captures, the output of Microsoft's "System Information" tool, and the current settings applied to my miniDSP  2x4 HD. 

The difference between config/output 1 and config/output 2 is whether or not exclusive mode is used. 
I was able to get sound output with the right combination of settings in exclusive mode, but the sound was very distorted. It sounded like very bad clipping, but it didn't get better after turning down every volume control I could find. When I played the Windows test tone at different sample rates, the pitch of the sound was (improperly) shifted up or down depending on whether I chose something below, at, or above the rate given in the camilladsp config file.

If I disable exclusive mode on either the playback or record side, camilladsp does not seem to ever enter its processing loop. The log output suggests it at least tries to start the processing thread, but nothing ever happens. The first time I run it after a reboot, it pauses briefly and then drops me back to the command prompt. Subsequent runs don't have the pause. 

Please let me know if I can provide more information. I am also interested in testing out Jack support under Windows. I didn't realize Jack had a Windows build until I checked today. It would be great to have that flexibility.
[settings_and_output.zip](https://github.com/HEnquist/camilladsp/files/6987003/settings_and_output.zip)

**Comment by HEnquist** [comment]
This seems quite odd. That shared mode config should work without any issues. It just hangs there until you kill it? Have you tried running with -vv? Won't help but may give more clues.

For the distortion in exclusive mode, it would really help to figure out if it comes from capture or playback.

A few ideas for things to try:
- Usb audio usually uses S24LE3, can you they that in exclusive mode on the minidsp?
- can you try with another capture source? For example the line-in of the built-in audio? Try with both shared and exclusive.
- same but instead with another playback device?

**Comment by nseyedtalebi** [comment]
I will try the things you suggested and report back.

It looks like it's crashing when it tries to build the I/O streams in shared mode:
```
PS C:\Users\NiMo3\camilladsp> .\camilladsp -vv .\config_2.yml
Aug 14 18:41:31.429 DEBG Read config file Some(".\\config_2.yml"), module: camilladsp
Aug 14 18:41:31.430 DEBG Config is valid, module: camilladsp
Aug 14 18:41:31.434 DEBG Wait for config, module: camilladsp
Aug 14 18:41:31.435 DEBG Config ready, module: camilladsp
Aug 14 18:41:31.435 DEBG Using channels [true, true], module: camilladsp
Aug 14 18:41:31.436 DEBG Build new pipeline, module: camillalib::filters
Aug 14 18:41:31.437 DEBG build filters, waiting to start processing loop, module: camillalib::processing
Aug 14 18:41:31.438 DEBG Using a playback channel capacity of 3 chunks., module: camillalib::wasapidevice
Aug 14 18:41:31.439 TRCE Build output stream, module: camillalib::wasapidevice
Aug 14 18:41:31.440 DEBG Using a capture channel capacity of 33 buffers., module: camillalib::wasapidevice
Aug 14 18:41:31.440 TRCE Build input stream, module: camillalib::wasapidevice
```
The following AppCrash event shows up in the Windows event log:
```
Faulting application name: camilladsp.exe, version: 0.0.0.0, time stamp: 0x611448ec
Faulting module name: camilladsp.exe, version: 0.0.0.0, time stamp: 0x611448ec
Exception code: 0xc0000005
Fault offset: 0x000000000025df6b
Faulting process id: 0x1084
Faulting application start time: 0x01d7915d87088ad9
Faulting application path: C:\Users\NiMo3\camilladsp\camilladsp.exe
Faulting module path: C:\Users\NiMo3\camilladsp\camilladsp.exe
Report Id: b6d3c393-1d6f-4c5d-85ed-1b2b21f97718
Faulting package full name: 
Faulting package-relative application ID: 
```

**Comment by HEnquist** [comment]
Ok I will have to look into that.
Could you also try the same with the published binaries for beta1 and beta2?

**Comment by HEnquist** [comment]
I can confirm that shared mode is broken in 0.6.0! 
It works in 0.6.0b2.

**Comment by HEnquist** [comment]
Please try version 0.6.1!
https://github.com/HEnquist/camilladsp/releases/tag/v0.6.1

**Comment by nseyedtalebi** [comment]
That seems to have done it!  Sorry for the delay, I've been behind on something for at least two years (or 33). 

I got clean sound out in both modes at least once. I also had some combination of things that gave me the distortion again, but I was just trying things quickly to see if the patch worked. If I can isolate the problem, I will open a separate issue on Github with details. Closing this issue out...
